source-pendo/source_pendo/models.py

```diff
@@ -0,0 +1,76 @@
+from estuary_cdk.capture.common import (
+    BaseDocument,
+    ResourceConfig,
+    ResourceState,
+)
+from estuary_cdk.capture.common import (
+    ConnectorState as GenericConnectorState,
+)
+from pydantic import AwareDatetime, BaseModel, Field
+
+# Supported resource types for the Pendo API that have an id and their corresponding resource name.
+RESOURCE_TYPES: list[tuple[str, str]] = [
+    ("feature", "Feature"),
+    ("guide", "Guide"),
+    ("page", "Page"),
+    ("report", "Report"),
+]
+
+# Supported metadata types for the Pendo API and their corresponding resource name.
+METADATA_TYPES: list[tuple[str, str]] = [
+    ("account", "AccountMetadata"),
+    ("visitor", "VisitorMetadata")
+]
+
+# Supported event types for the Pendo Aggregation API, their corresponding resource name, their associated identifying field, and their "last updated" timestamp field.
+EVENT_TYPES: list[tuple[str, str, str]] = [
+    ("pageEvents", "PageEvents", "/pageId", "lastTime"),
+    ("featureEvents", "FeatureEvents", "/featureId", "lastTime"),
+    ("trackEvents", "TrackEvents", "/trackTypeId", "lastTime"),
+    ("guideEvents", "GuideEvents", "/guideId", "guideTimestamp"),
+    ("pollEvents", "PollEvents", "/pollId", "guideTimestamp")
+]
+
+
+class EndpointConfig(BaseModel):
+    startDate: str = Field(
+        description="UTC date and time in the format YYYY-MM-DDTHH:MM:SSZ. Any event data generated before this date will not be replicated.",
+        title="Start Date",
+        examples= ["2024-08-27T00:00:00Z"],
+        pattern= "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$",
+        format= "date-time",
+    )
+    apiToken: str = Field(
+        description="API Token to use for Authentication",
+        title="API Token",
+        json_schema_extra={"secret": True},
+    )
+
+
+ConnectorState = GenericConnectorState[ResourceState]
+
+
+class Resource(BaseDocument, extra="allow"):
+    id: str = ""
```

Comment:
There are a lot of fields in these models that have default values. Usually they shouldn't, since that is actually schematized with a JSON default. Especially for `id` which I'd imagine we always want to have present, and it should be an error if it isn't there. And for `id` in particular setting a default value makes it's schema not have it be a required field, and having the collection key be a require field is pretty much universally required in order for the collection to be materialized.

So I'd generally recommend getting rid of the default values for these models. You _could_ use the `| None` union to represent optional fields, but probably best for those to just not even explicitly schematize them.

Reply:
For the various event streams, I use the `| None` union for fields that exist on one stream, but don't exist on other streams (ex: `pageId` only exists on `pageEvents` but not `featureEvents`). To avoid having those fields be optional for the other streams, I think I'll have to make separate models for each one. I'll give it a try.

Reply:
You could consider just omitting the fields entirely too since they will be determined by schema inference. Having one model that only has the fields that are common to all event streams would be totally fine.

Reply:
I ended up having separate models for each of the event streams - each use a different field as a key. And no optional or default values for any models.